src/common/fundFromTurbo.ts

```diff
@@ -0,0 +1,252 @@
+  paymentUrl?: string;
+  // The logger to use
+  logger?: ILogger;
+  // The HTTP client to use
+  axios?: AxiosInstance;
+  signer?: FundFromTurboSigner;
+}
+
+export type InitiateArNSPurchaseParams = {
+  type?: string;
+  years?: number;
+  increaseQty?: number;
+  name: string;
+  intent: Intent;
+  processId?: TransactionId;
+};
+
+export async function signedRequestHeadersFromSigner(
+  signer: FundFromTurboSigner,
+  nonce: string = v4(),
+): Promise<RawAxiosRequestHeaders> {
+  await (signer as ArconnectSigner).setPublicKey?.();
+  const signature = await signer.sign(Uint8Array.from(Buffer.from(nonce)));
+
+  let publicKey: string;
+  switch (signer.signatureType) {
+    case SignatureConfig.ARWEAVE:
+      publicKey = toB64Url(signer.publicKey);
+      break;
+    case SignatureConfig.ETHEREUM:
+      publicKey = '0x' + signer.publicKey.toString('hex');
+      break;
+    // TODO: solana sig support
+    // case SignatureConfig.SOLANA:
+    // case SignatureConfig.ED25519:
+    default:
+      throw new Error(
+        `Unsupported signer type for signing requests: ${signer.signatureType}`,
+      );
+  }
+
+  return {
+    'x-public-key': publicKey,
+    'x-nonce': nonce,
+    'x-signature': toB64Url(Buffer.from(signature)),
+    'x-signature-type': signer.signatureType,
+  };
+}
+
+export type ArNSPurchaseReceipt = InitiateArNSPurchaseParams & {
+  wincQty: string;
+  mARIOQty: string;
+  usdArRate: number;
+  createdDate: string;
+};
+
+export type ArNSPurchaseResult = {
+  arioWriteResult: AoMessageResult;
+  purchaseReceipt: ArNSPurchaseReceipt;
+};
+
+export interface FundFromTurboInterface {
+  getArNSPurchasePrice(params: InitiateArNSPurchaseParams): Promise<number>;
+  initiateArNSPurchase(
+    params: InitiateArNSPurchaseParams,
+    options: WriteOptions,
+  ): Promise<ArNSPurchaseResult>;
+  getArNSPurchaseReceipt(intent: string): Promise<ArNSPurchaseReceipt>;
+}
+
+export class FundFromTurbo {
+  private readonly paymentUrl: string;
+  private readonly axios: AxiosInstance;
+  private readonly logger: ILogger;
+  private readonly signer?: FundFromTurboSigner;
+
+  constructor({
+    paymentUrl = 'https://payment.ardrive.io',
+    axios = createAxiosInstance(),
+    logger = Logger.default,
+    signer,
+  }: TurboConfig) {
+    this.paymentUrl = paymentUrl;
+    this.axios = axios;
+    this.logger = logger;
+    this.signer = signer;
+  }
+
+  public async getArNSPurchasePrice({
+    intent,
+    name,
+    increaseQty,
+    type,
+    years,
+  }: InitiateArNSPurchaseParams): Promise<{ winc: string; mARIO: mARIOToken }> {
+    const url = new URL(`${this.paymentUrl}/v1/arns/price/${intent}/${name}`);
+    if (increaseQty !== undefined) {
+      url.searchParams.append('increaseQty', increaseQty.toString());
+    }
+    if (type !== undefined) {
+      url.searchParams.append('type', type);
+    }
+    if (years !== undefined) {
+      url.searchParams.append('years', years.toString());
+    }
```

Comment:
I think this is fine since its a small amount of params, but we could also do something like:

```suggestion
url.search = new URLSearchParams(
  Object.fromEntries(
    Object.entries({increaseQty, type, years}).filter(item => !!item)
      )
    ).toString()
```

URLSearchParams.toString() should handle numbers, strings, and other primitive with toString() methods appropriately.

Reply:
harder to read at first glance. I think something like that would read better and be pretty useful as a utility function like `urlSearchParamsFromOptionalInputs`